### ISSUE TITLE
Polling from Pool

### ISSUE THREAD

**Issue by LJBD** [sardana-org/sardana#456]
Hello!

TL;DR: too much polling causes the Pool to crash. Questions below.

Long version:
This is a long report regarding what we've experienced at SOLARIS working with Sardana 2.2.2 over the last 2 weeks. Since we're running CentOS 7, we launch Pool and MacroServer separately.

We were introducing HDB++ archiving to one of our beamlines which has just entered commissioning phase. The beamline's manager wanted to have 6 attributes per axis archived with intervals ranging from 3 to 10 seconds - nothing too fancy. Pool for that beamline controls 55 motors, 4 of which are dummies. We configured polling (and archive events, too) accordingly, only for *Position* attributes going for a polling period of 1 s.

Unfortunately, it turned out that Pool cannot handle that - it crashed approximately once per 1.5 - 2 hours. We tried to do several things to prevent that:
1. First, we tried adding polling threads for Pool and balance the load between them. It made the DS to crash even more frequently, so we went back to a single thread very quickly.
2. Second, we tried disabling the archiving - it yielded no results.
3. Third, we increased polling periods for all attributes apart from *Position*. This stopped messages like "polling thread is late" when one subscribed to an attribute through PyTango, but it didn't prevent the Pool from crashing one more time.
4. Fourth, we disabled the polling completely and turned the archiving off, because the beamline's manager started to become impatient. This solved the issue.

What is also interesting, from a user's side we saw symptoms of the Pool not being able to handle such amounts of polling as well (aside from the obvious effects of a crashed Pool). There was an error when performing movements with physical motors (pseudo-motors are not affected). It didn't happen during every movement nor did it happen with every motor, but often we got a message like the ones you can see on the attached pictures when trying to run macros through spock.
When it happened during a movement macro (like umv), the macro stopped, but the motor continued its movement. 

![m1roll-error-01](https://cloud.githubusercontent.com/assets/9516795/25218076/d28c5cfa-25a9-11e7-88d3-d897ff0788f9.png)

![m1yaw-error-01](https://cloud.githubusercontent.com/assets/9516795/25218075/d28b5a9e-25a9-11e7-8089-31b1209a3269.png)

When it happened during a scan, it aborted with an error: "Motion ended with None".
![sardana-error-02](https://cloud.githubusercontent.com/assets/9516795/25218074/d28b4216-25a9-11e7-867d-86a0324823e7.png)

As a side note, this *TypeError* regarding TaurusDevState in a State of a motor looks like a bug.

Since the archiving is an important service, we were wondering about possible solutions and came up with one: divide the Pool into several smaller ones.

We have several questions regarding this situation:
1. How many motors do you have on a Pool?
2. Do you have many Pools per beamline or just one?
3. Are there any disadvantages of dividing a Pool?
4. Does anyone use HDB++ with Sardana?
5. Can you tell from inside Sardana that a Pool doesn't work properly (e.g. its polling thread is late)?
6. Can the message *"Motion ended with/in None"* be a bit more specific?

Best regards,
Łukasz Dudek

**Comment by cmft** [comment]
Hi @LJBD,

Could you give us more information about your setup (tango, PyTango, and taurus version)? 

> How many motors do you have on a Pool?
We have BLs with more than 50 physical motors (~70).
How many motor controllers do you have?

> Do you have many Pools per beamline or just one?
Yes, we do. Normally we have two Pools (for safety reason)

> Are there any disadvantages of dividing a Pool?
No as far I know. 

>  Does anyone use HDB++ with Sardana?
We only have one BL with HDB++  but we don't use with Sardana. I don't remember any trouble with minimum polling time but better ask it a HDB++ expert.

>  Can you tell from inside Sardana that a Pool doesn't work properly (e.g. its polling thread is late)? We should report it in the log.

>  Can the message "Motion ended with/in None" be a bit more specific?
This message is reported by the GScan and None (is the state of the moveable) and I can image that it is None due to the  polling thread problems.

Did you see archiving problems in any other device? Do you believe that the test case can be simplified?

Do you know the status of the crashed motor in icepapcms?
I saw time to time crashed in the Pool of my BeamLine when pyicepap raises an exception while it is moving (and the icepap controller propagate it to the Pool). Did you see any exception in the logs?

Maybe a faster polling (of the same attributes) can cause the same behaviour in a Pool with only one Axis.


Best,
Carlos

**Comment by reszelaz** [comment]
The behavior of crashing the Pool when the polling is active is of course not normal. We will work on fixing it but we will need your cooperation. Apart of @cmft questions, I have some more.

> Unfortunately, it turned out that Pool cannot handle that - it crashed approximately once per 1.5 - 2 hours.

1. Does this crash also happen when there is no activity on the Pool i.e. no motion, acquisition in progress, etc. and just the Pool process running (you could even try to close all the guis that may execute readouts).

2. Did you find a way of reproducing it more frequently? Making the polling more frequent, adding more attributes to the polling? Have you seen any relation with the crash frequency when trying different configurations? 

3. Could you attach the backtrace from the corefile? And the last logs of the Pool just before the crash?

And now I will try to answer your questions:

> How many motors do you have on a Pool?

This depends, but I would say that between 30 and 80 per beamlines.

> Do you have many Pools per beamline or just one?

In general, we have only one Pool for the elements of the beamline. We have another one for the front end (it is questionable if this is already part of the beamline or it belongs to the machine) but this is because of the necessity to control its elements from the machine control room and the network topology constraints.. and not of the problems like you describe. Sporadically we add a temporary extra Pool when we need to do some ad-hoc operation at the beamline e.g. commissioning of some new instruments..

> Are there any disadvantages of dividing a Pool?

Well, I would say no, but... For sure there are some thing that you should keep in mind when you decide to redistribute the elements along the pools e.g. the pseudo motors and motor groups can be based only on the physical motors belonging to the same pool, measurement groups can use only channels belonging to the same pool or the software synchronization in the position domain can be achieved only on the motor belonging to the same pool.

> Does anyone use HDB++ with Sardana?

I think that at Alba not yet. At some beamlines we archive some Pool values but with the “old” archiving system. AFAIK the archiving frequency is not so high that you try to achieve with the HDB++. Here @sergirubio and my other colleagues could give you more details. 

> Can you tell from inside Sardana that a Pool doesn't work properly (e.g. its polling thread is late)?

I’m not aware about any known problems related to that. But we don’t use polling with Pool. We rely on the change events pushed on write and during the actions.

> Can the message "Motion ended with/in None" be a bit more specific?

This is certainly a bug, it would be nice if you could find a way to reproduce it and make a specific bug report. I imagine that it won’t be hard to fix it.

**Comment by LJBD** [comment]
Thanks for your quick reactions!

I'll try to answer everything.

> Could you give us more information about your setup (tango, PyTango, and taurus version)?

We use Tango 9.2.2, PyTango 9.2.0 and Taurus 4.0.1 for this beamline.

> Did you see archiving problems in any other device?

Well, we haven't had problems with archiving. However we've also seen change events from Lima 1.7 being bunched (as in: no events for 10 s, then 4 events at once), but it appeared only one client machine.

> Do you believe that the test case can be simplified?

I suppose so, I'll try reproducing the problem on a development instance of Sardana and then maybe in a CentOS Docker container as well, if I have time. I don't think that I'll manage to do so before 10 May, though, but I'll let you know as soon as I've come up with anything interesting.

> Do you know the status of the crashed motor in icepapcms?

AFAIK they were operational all the time. I haven't seen any IcePAP-related errors in the log prior to a crash apart from that Sardana can't get *PosIndexer* and *EncIndexer* attributes.

> Does this crash also happen when there is no activity on the Pool i.e. no motion, acquisition in progress, etc. and just the Pool process running (you could even try to close all the guis that may execute readouts).

We didn't test it without client applications and right now there's ongoing commissioning, so we can't do anything. We have a shutdown from 8 May and we may be able to do some further tests then.

> Did you find a way of reproducing it more frequently? Making the polling more frequent, adding more attributes to the polling? Have you seen any relation with the crash frequency when trying different configurations?

That's not what we had in mind then, but it's a very interesting point. Again, after 8 May we should be able to test that, too.

> Could you attach the backtrace from the corefile? And the last logs of the Pool just before the crash?

Unfortunately, it turns out that ABRT has run into some problems when creating core dump and it hasn't been saved. The logs are also gone - I wasn't aware of the fact that they are rotated in 6 files and for our Sardana with log level set to *DEBUG* those 6 files are filled in around 6 hours...
I've only got a screenshot of "Standard Error" from Astor, attached below. Pool logs stop before this line:
```console
*** Error in `/usr/bin/python2': double free or corruption (...)
```

![image](https://cloud.githubusercontent.com/assets/9516795/25429732/37a6e446-2a7a-11e7-8355-5e47fb77627c.png)

Before further tests I will solve problems with ABRT and hopefully I will be able to come back with a corefile.

> (it is questionable if this is already part of the beamline or it belongs to the machine) 

I guess that everyone has had discussions about that :)

> But we don’t use polling with Pool.

AFAIK if polling is enabled and its thread is late, you get *API_EventTimeout* errors also when subscribing to change events, so you should be able to see that inside Pool, too.

> This is certainly a bug, it would be nice if you could find a way to reproduce it and make a specific bug report. I imagine that it won’t be hard to fix it.

Okay, I will, but again, possibly after 10 May (or maybe I'll try to find some colleague to delegate it to).

**Comment by reszelaz** [comment]
Just one hint, you cound change the file logging setting in the [sardanacustomsettings.py](https://github.com/sardana-org/sardana/blob/develop/src/sardana/sardanacustomsettings.py#L42) so that you either increse the size of one log file or the number of log files.

One can see in the backtrace that you attached that the crash must be somehow related with the [Tango::Attribute::fire_change_event](http://www.esrf.eu/computing/cs/tango/tango_doc/kernel_doc/cpp_doc/classTango_1_1Attribute.html#ae023edda1e0bf70d5fda7c3dc8160351). I would also not ignore the `posindexer` and `encindexer` errors. Do you have a clue why they happen?

**Comment by LJBD** [comment]
We have managed to conduct some more tests lately and here are the results (given in no particular order):

1. We've changed a VM on which Sardana device servers (1 Pool and 3 MacroServers) are running.
2. We haven't been able to get ABRT to work. Here is the error that shows up when a process ends in an unusual way: 
```console
Apr 24 16:35:26 sar abrt-hook-ccpp: Process 18370 (python2.7) of user 501 killed by SIGABRT - dumping core
Apr 24 16:36:17 sar abrt-hook-ccpp: Failed to create core_backtrace: dwfl_linux_proc_report failed: no error
Apr 24 16:36:18 sar abrt-server: Interpreter crashed, but no packaged script detected: '/usr/bin/python2 /usr/bin//MacroServer macroserver-bl04bm'
Apr 24 16:36:18 sar abrt-server: 'post-create' on '/var/spool/abrt/ccpp-2017-04-24-16:35:26-18370' exited with 1
Apr 24 16:36:18 sar abrt-server: Deleting problem directory '/var/spool/abrt/ccpp-2017-04-24-16:35:26-18370'
```
We haven't been able to find any real solution to this problem on the Internet.
3. On the plus side, we haven't been able to crash Sardana any more.
4. We've made a series of tests where we enabled polling for 3 attributes per motor for all of them (including simulated ones). For both 3 and 10 seconds of polling period we've seen those *API_EventTimeout* (errors caused by the polling thread being late) for several motors including the simulated ones - this might indicate that it wasn't an issue with hardware communication.
5. We've tried again adding polling threads to the Pool Device Server. This time it worked as expected: after assigning each motor its individual polling thread, we haven't seen any *API_EventTimeout* for the last several days. Right now polling periods are 1 s for 3 attributes per motor and we have 57 polling threads.
6. We've tested how small a period can we set in order not to get those polling errors. It turns out that at 250 ms readings are still stable and no errors are observed.
7. We've enabled HDB++ archiving and configured some very long scans. For the last day it has been stable, no further errors were observed.

My guess is that there was some problem with a VM that we initially used to launch Sardana.

**Comment by reszelaz** [comment]
It is good to hear that you do not suffer from this problem anymore. Let's wait a couple of days more with closing this issue, just in case the problem appears again.

Thanks for performing thesee tests and confirming that Sardana works with HDB++.

I don't have experience with configuration of the ABRT so I won'the help hear, sorry...

**Comment by LJBD** [comment]
We haven't had any problems so far, so I'm closing this issue.
